File: clients/cpp/model/Specialty.cpp
#include "Specialty.hpp"
#include <stdexcept>

namespace model {

// Read Specialty from input stream
Specialty readSpecialty(InputStream& stream) {
    switch (stream.readInt()) {
    case 0:
        return Specialty::LOGISTICS;
    case 1:
        return Specialty::PRODUCTION;
    case 2:
        return Specialty::COMBAT;
    default:
        throw std::runtime_error("Unexpected tag value");
    }
}

// Get string representation of Specialty
std::string specialtyToString(Specialty value) {
    switch (value) {
    case Specialty::LOGISTICS:
        return "LOGISTICS";
    case Specialty::PRODUCTION:
        return "PRODUCTION";
    case Specialty::COMBAT:
        return "COMBAT";
    default:
        throw std::runtime_error("Impossible happened");
    }
}

}

File: clients/scala/src/main/scala/spb_ai_champ/model/Specialty.scala
package spb_ai_champ.model

import spb_ai_champ.util.StreamUtil

/**
 * Player's specialty
 */
sealed abstract class Specialty (val tag: Int) {
    /**
     * Write Specialty to output stream
     */
    def writeTo(stream: java.io.OutputStream) {
        StreamUtil.writeInt(stream, tag)
    }
}

object Specialty {
    /**
     * Logistics. Increased travel distance
     */
    case object LOGISTICS extends Specialty(0)
    /**
     * Production. Increased work speed
     */
    case object PRODUCTION extends Specialty(1)
    /**
     * Combat. Increased damage
     */
    case object COMBAT extends Specialty(2)

    /**
     * Read Specialty from input stream
     */
    def readFrom(stream: java.io.InputStream): Specialty =
        StreamUtil.readInt(stream) match {
            case LOGISTICS.tag => LOGISTICS
            case PRODUCTION.tag => PRODUCTION
            case COMBAT.tag => COMBAT
            case _ => throw new java.io.IOException("Unexpected tag value")
        }
}